### PR TITLE
refactor: split GGUF quantization into hardware backends and schemes

### PR DESCRIPTION
**Pull request by LProDreamAll**
## Motivation

Add schemes to GGUF instead of storing all classes in a single file, and split kernel call and weight init. Follow up to the quantization refactor roadmap: #15194.

## Modifications

Refactored GGUF to align with the scheme-based quantization structure used by modelslim and compressed_tensors.

Moved GGUF implementations out of the monolithic `quantization/gguf.py` into `quantization/gguf/`, with scheme implementations under `quantization/gguf/schemes/`.

Added `get_linear_scheme`, `get_embedding_scheme`, and `get_moe_scheme` to `gguf/gguf.py` for explicit scheme selection.

Unified GGUF quant methods into thin wrappers that delegate to `layer.scheme`.

Split backend-specific kernel logic into:

* `hardware_backend/gpu/quantization/gguf_kernels.py`
* `hardware_backend/npu/quantization/gguf_kernels.py`

## Accuracy Tests

GPU test: `python3 test/registered/quant/test_gguf.py`

- Model: `Qwen/Qwen2-1.5B-Instruct-GGUF` (`q4_k_m`)
- Result: passed locally

## Benchmarking and Profiling

N/A — behavior-preserving refactor, no expected perf change.

## Checklist

- [ ] Format your code according to the Format code with pre-commit.
- [ ] Add unit tests according to the Run and add unit tests.
- [ ] Update documentation according to Write documentations.
- [ ] Provide accuracy and speed benchmark results according to Test the accuracy and Benchmark the speed.
- [ ] Follow the SGLang code style guidance.

---
### CI States

Latest PR Test (Base): :x: [Run #26759288403](https://github.com/sgl-project/sglang/actions/runs/26759288403)
Latest PR Test (Extra): :x: [Run #26759287002](https://github.com/sgl-project/sglang/actions/runs/26759287002)

### CODE REVIEW

**Comment by LProDreamAll** [comment]
@BBuf @AniZpZ @ch-wan Could one of you help trigger CI by posting `/tag-run-ci-label`? This PR is a follow-up to the quantization refactor roadmap (#15194), splitting GGUF into hardware backends and schemes. CI was skipped because the `run-ci` label is missing.

**Comment by LProDreamAll** [comment]
@TamirBaydasov    Can you help me?

**Comment by TamirBaydasov** [comment]
> @TamirBaydasov Can you help me?

Hi, I will review the PR this week. But I can't add labels, so I will ask as well.

**Comment by LProDreamAll** [comment]
@TamirBaydasov  Thanks for confirming the review! I am readily available at any time.

**Comment by TamirBaydasov** [comment]
LGTM. @OrangeRedeng pls take a look as well if you have time

File: python/sglang/srt/hardware_backend/npu/quantization/gguf_kernels.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+class GGUFAscendEmbeddingKernel(GGUFAscendLinearKernel):
+    def embedding(self, layer: torch.nn.Module, x: torch.Tensor) -> torch.Tensor:
+        return torch.embedding(layer.dequantized_weight, x)
+
+
+class GGUFAscendMoEKernel:
+    def __init__(self, quant_config: Optional["QuantizationConfig"] = None):
+        self.quant_config = quant_config
+        self.moe_runner_config: Optional["MoeRunnerConfig"] = None
+        self.params_dtype: torch.dtype | None = None
```
Comment (TamirBaydasov) [inline python/sglang/srt/hardware_backend/npu/quantization/gguf_kernels.py:150]: Could you please look at the possibility of adding a `self.kernel` field here that is going to be used in `process_weights_after_loading` and `apply`?
Example:
https://github.com/sgl-project/sglang/pull/21126/changes#diff-fa915d91e4635d2513e4f10acd16d80cfa9d5a08ca46bc8fb52f98cedde5bd5cR74

File: python/sglang/srt/hardware_backend/npu/quantization/gguf_kernels.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    dequant_np = gguf_dequantize(qweight_cpu, qweight_type)
+    result = torch.from_numpy(dequant_np).to(dtype=dtype, device=qweight.device)
+    result = result.reshape(rows, cols)
+    return result
+
+
+class GGUFAscendLinearKernel:
+    def __init__(self, quant_config: Optional["QuantizationConfig"] = None):
+        self.quant_config = quant_config
+        self.params_dtype: torch.dtype | None = None
```
Comment (TamirBaydasov) [inline python/sglang/srt/hardware_backend/npu/quantization/gguf_kernels.py:70]: Same question as in: https://github.com/sgl-project/sglang/pull/26942/changes#r3474859621

**Comment by TamirBaydasov** [comment]
Please, update accuracy tests with NPU models as well. Good measurement is gsm8k dataset.
